sstables/sstables.cc

```diff
@@ -1097,28 +1097,32 @@ future<> sstable::read_simple(T& component, const io_priority_class& pc) {
 
     auto file_path = filename(Type);
     sstlog.debug(("Reading " + sstable_version_constants::get_component_map(_version).at(Type) + " file {} ").c_str(), file_path);
-    return new_sstable_component_file(_read_error_handler, Type, open_flags::ro).then([this, &component] (file fi) {
-        auto fut = fi.size();
-        return fut.then([this, &component, fi = std::move(fi)] (uint64_t size) {
-            auto r = make_lw_shared<file_random_access_reader>(std::move(fi), size, sstable_buffer_size);
-            auto fut = parse(*_schema, _version, *r, component);
-            return fut.finally([r] {
-                return r->close();
-            }).then([r] {});
-        });
-    }).then_wrapped([file_path] (future<> f) {
         try {
-            f.get();
-        } catch (std::system_error& e) {
+
+        file fi = co_await new_sstable_component_file(_read_error_handler, Type, open_flags::ro);
+        uint64_t size = co_await fi.size();
+
+        auto r = file_random_access_reader(std::move(fi), size, sstable_buffer_size);
+        std::exception_ptr eptr;
+        try {
+            co_await parse(*_schema, _version, r, component);
+        } catch (...) {
+            eptr = std::current_exception();
+        }
+        co_await r.close();
+
+        if (eptr) {
+            std::rethrow_exception(std::move(eptr));
+        }
+
+        }  catch (std::system_error& e) {
             if (e.code() == std::error_code(ENOENT, std::system_category())) {
-                return make_exception_future<>(malformed_sstable_exception(file_path + ": file not found"));
+                throw malformed_sstable_exception(file_path + ": file not found");
```

Comment:
Since now it's a coroutine, maybe the `co_await coroutine::return_exception()` should be used?

Reply:
@raphaelsc , this suggestion is still relevant from my side :)

Reply:
> @raphaelsc , this suggestion is still relevant from my side :)

IIRC, we cannot use co_await inside the catch statement. So we'd need all that verbose idiom of saving exception and maybe rethrowing it post catch. I took a look at raft/ and noticed they're throwing directly in coroutine frames, inside the catch statements... The only difference, according to the doc, is that the coroutine::exception reduces one hop by propagating the exception into the coroutine directly, but for the system_error and malformed_sstable_exception, which are very rare anyway, we can just use throw for gaining readability. What do you think?